criu/pidfd.c

```diff
@@ -269,32 +266,31 @@ static int collect_one_pidfd(void *obj, ProtobufCMessage *msg, struct cr_img *i)
 	info->pidfe = pb_msg(msg, PidfdEntry);
 	pr_info_pidfd("Collected ", info->pidfe);
 
+	info->dead = NULL;
 	if (info->pidfe->nspid != -1)
 		goto out;
 
 	dead = lookup_dead_pidfd(info->pidfe->ino);
-	if (dead) {
-		mutex_lock(&dead->pidfd_lock);
-		dead->count++;
-		mutex_unlock(&dead->pidfd_lock);
-		goto out;
-	}
-
-	dead = shmalloc(sizeof(*dead));
 	if (!dead) {
-		pr_err("Could not allocate shared memory..\n");
-		return -1;
+		dead = xmalloc(sizeof(*dead));
+		if (!dead) {
+			pr_err("Could not allocate memory..\n");
+			return -1;
+		}
+
+		INIT_HLIST_NODE(&dead->hash);
+		dead->list = NULL;
+		dead->ino = info->pidfe->ino;
+		dead->creator_id = info->pidfe->id;
+		hlist_add_head(&dead->hash, &dead_pidfd_hash[dead->ino % DEAD_PIDFD_HASH_SIZE]);
 	}
 
-	INIT_HLIST_NODE(&dead->hash);
-	dead->ino = info->pidfe->ino;
-	dead->count = 1;
-	dead->pid = -1;
-	mutex_init(&dead->pidfd_lock);
+	info->dead = dead;
+	info->next = dead->list;
+	dead->list = info;
+	if (dead->creator_id > info->pidfe->id)
```

Comment:
We have discussed that with @avagin and a tricky thing here is that we need to ensure that there are no deadlock when process A waits on process B to restore some fd1, while process B waits on process A to restore (another) fd2. To prevent those deadlocks we have special way of deciding who will restore an fd (see `fdinfo_rst_prio`) and `mark_pipe_master_cb` (good example). I think we need to implement something like that here.

Reply:
@bsach64 if you want I can do this on top of this PR so you can squash it later.

Reply:
Okay Alex!